ecmascript/dep-graph/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+use swc_common::{
+    comments::{Comment, SingleThreadedComments},
+    Loc, DUMMY_SP,
+};
+pub use swc_common::{SourceMap, Span};
+use swc_ecma_ast as ast;
+use swc_ecma_visit::{self, Node, Visit, VisitWith};
+
+pub fn analyze_dependencies(
+    module: &ast::Module,
+    source_map: &SourceMap,
+    comments: &SingleThreadedComments,
+) -> Vec<DependencyDescriptor> {
+    let mut v = DependencyCollector {
+        comments,
+        source_map,
+        items: vec![],
+    };
+    module.visit_with(&ast::Invalid { span: DUMMY_SP }, &mut v);
+    v.items
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum DependencyKind {
+    Import,
+    ImportType,
+    Export,
+    ExportType,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DependencyDescriptor {
+    pub kind: DependencyKind,
+    /// A flag indicating if the import is dynamic or not.
+    pub is_dynamic: bool,
```

Comment:
Debatable - maybe better to use separate kind for dynamic imports as well as `require()`?

Reply:
I think flag is right way to go as `require`  can be both dynamic or static.
swc_bundler actually handle requires as a static library if they are called at top level.
e.g.
```js
const { join } = require('path');
```
This is an unconditional import with specifier 'join'.

Reply:
Right, so I guess we can just add another `DependencyKind::Require` to handle them?

Reply:
Yes, it will work.

Reply:
Done